climada/hazard/base.py

```diff
@@ -651,9 +651,29 @@ def from_raster_xarray(
         if not isinstance(data, xr.Dataset):
             LOGGER.info("Loading Hazard from file: %s", data)
             data: xr.Dataset = xr.open_dataset(data, chunks="auto")
+            data_is_mine = True
         else:
             LOGGER.info("Loading Hazard from xarray Dataset")
+            data_is_mine = False
+        try:
+            return cls._from_raster_xarray_dataset(data, hazard_type,
+                intensity_unit, intensity, coordinate_vars, data_vars, crs, rechunk)
+        finally:
+            if data_is_mine:
+                data.close()
```

Comment:
```suggestion
        if not isinstance(data, xr.Dataset):
            LOGGER.info("Loading Hazard from file: %s", data)
            with xr.open_dataset(data, chunks="auto") as dataset:
                return cls._from_raster_xarray_dataset(dataset, hazard_type,
                    intensity_unit, intensity, coordinate_vars, data_vars, crs, rechunk)
        else:
            return cls._from_raster_xarray_dataset(data, hazard_type,
                intensity_unit, intensity, coordinate_vars, data_vars, crs, rechunk)
```

Reply:
Indeed. This approach has the advantage that it works the same way with both types of inputs. But it has the following disadvantages, in my opinion:
* It defines a "private" classmethod. As classmethods instantiate objects, they should never be marked as "private".
* Although "private", this classmethod needs to define all parameters and contain the docstring, such that implementation and documentation are not separated.
* The documentation has to either be doubled in the public classmethod, or its docs must refer to the private method that actually indicates users not to touch it

I would rather go for the two-method approach. Users then must take the somewhat concious decision to either open a file themselves or let Climada do that, and choose one of the methods accordingly.

Reply:
I'm not opposed to the two-method approach, but I can't quite follow the argumentation:

- why should classmethods never be "private" (in this case, "protected" strictly speaking)?
- I wouldn't want to have a docstring for any "protected" method. They're not supposed to be used by anyone but the package developers and hence shouldn't show up in the API reference. But adding a docstring leads to just that: the method is listed even though it's "protected".
- with the two-methods approach the documentation is _de facto_ doubled - so there is nothing to gain even if we did add a docstring to the "protected" method (which we shouldn't imo, s.a.)?

Reply:
> why should classmethods never be "private"

A classmethod indicates a specialized "constructor" for an object. "Private" or "protected" means that the method should only be called by the class (or derived classes) itself. Why should a constructor of an object only be called by the object itself? This seems like an anti-pattern to me 🤔 

> I wouldn't want to have a docstring for any "protected" method.

That's exactly my point. You have the doctring in the "public" method, but all the implementation is actually in the "protected" method, which does not have a docstring. This separates docs from code, which I find dangerous.

> with the two-methods approach the documentation is de facto doubled

No, you just link to the other public method. Compare, e.g., [`xr.load_dataset`](https://docs.xarray.dev/en/stable/generated/xarray.load_dataset.html) and [`xr.open_dataset`](https://docs.xarray.dev/en/stable/generated/xarray.open_dataset.html), I would do it exactly like that.

> I'm not opposed to the two-method approach

Shall I come up with a proposal? What do you think about the method names?

Reply:
> Why should a constructor of an object only be called by the object itself?
To me it's still a pattern since, actually, it's called by the class and not the object.

> This separates docs from code, which I find dangerous.
Despite the danger I prefer stable docs for the exposed interface alone (exposed as in API reference). How the stuff is dealt with internally must not bother the user and developers are welcome to change it without much fuzz and without fear of disruption.
Public documentation of private methods is kind of corrupting the whole idea of privacy. 😉
Besides, there are dozens of protected methods in CLIMADA without a docstring, so this would not coma as a novelty.

Having said that - yes, go ahead. 😃

About the names I'm pretty indifferent, however I prefer `*_xarray_*` over `*_xr_*` and if it's my choice I place `xarray` before `raster`, like `from_xarray_raster_file`. Your call.